File: temp_convert.go
// temp_convert.go - Конвертер температур на Go (CLI)
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Scale struct {
	ToKelvin   func(float64) float64
	FromKelvin func(float64) float64
	Name       string
	Symbol     string
}

var scales = map[string]Scale{
	"c": {
		ToKelvin:   func(x float64) float64 { return x + 273.15 },
		FromKelvin: func(x float64) float64 { return x - 273.15 },
		Name:       "Цельсий", Symbol: "°C",
	},
	"f": {
		ToKelvin:   func(x float64) float64 { return (x-32)*5/9 + 273.15 },
		FromKelvin: func(x float64) float64 { return (x-273.15)*9/5 + 32 },
		Name:       "Фаренгейт", Symbol: "°F",
	},
	"k": {
		ToKelvin:   func(x float64) float64 { return x },
		FromKelvin: func(x float64) float64 { return x },
		Name:       "Кельвин", Symbol: "K",
	},
	"ra": {
		ToKelvin:   func(x float64) float64 { return x * 5 / 9 },
		FromKelvin: func(x float64) float64 { return x * 9 / 5 },
		Name:       "Ранкин", Symbol: "°Ra",
	},
	"re": {
		ToKelvin:   func(x float64) float64 { return x*5/4 + 273.15 },
		FromKelvin: func(x float64) float64 { return (x - 273.15) * 4 / 5 },
		Name:       "Реомюр", Symbol: "°Ré",
	},
	"n": {
		ToKelvin:   func(x float64) float64 { return x*100/33 + 273.15 },
		FromKelvin: func(x float64) float64 { return (x - 273.15) * 33 / 100 },
		Name:       "Ньютон", Symbol: "°N",
	},
	"de": {
		ToKelvin:   func(x float64) float64 { return 373.15 - x*2/3 },
		FromKelvin: func(x float64) float64 { return (373.15 - x) * 3 / 2 },
		Name:       "Делиль", Symbol: "°De",
	},
	"ro": {
		ToKelvin:   func(x float64) float64 { return (x-7.5)*40/21 + 273.15 },
		FromKelvin: func(x float64) float64 { return (x-273.15)*21/40 + 7.5 },
		Name:       "Рёмер", Symbol: "°Rø",
	},
}

func convert(value float64, from, to string) (float64, error) {
	if from == to {
		return value, nil
	}
	f, ok := scales[from]
	if !ok {
		return 0, fmt.Errorf("неизвестная шкала: %s", from)
	}
	t, ok := scales[to]
	if !ok {
		return 0, fmt.Errorf("неизвестная шкала: %s", to)
	}
	kelvin := f.ToKelvin(value)
	return t.FromKelvin(kelvin), nil
}

func main() {
	var (
		value     float64
		from      string
		to        string
		precision int
		batch     string
		output    string
		list      bool
		rangeArgs string
	)
	flag.Float64Var(&value, "value", 0, "Значение для конвертации")
	flag.StringVar(&from, "from", "c", "Исходная шкала")
	flag.StringVar(&to, "to", "f", "Целевая шкала")
	flag.IntVar(&precision, "precision", 2, "Количество знаков после запятой")
	flag.StringVar(&batch, "batch", "", "Файл со значениями (по одному на строку)")
	flag.StringVar(&output, "output", "", "Файл для сохранения результатов")
	flag.BoolVar(&list, "list", false, "Показать список шкал")
	flag.StringVar(&rangeArgs, "range", "", "Диапазон для таблицы: start,end,step")
	flag.Parse()

	if list {
		fmt.Println("Доступные шкалы:")
		for key, s := range scales {
			fmt.Printf("  %s: %s (%s)\n", key, s.Name, s.Symbol)
		}
		return
	}

	if rangeArgs != "" {
		parts := strings.Split(rangeArgs, ",")
		if len(parts) != 3 {
			fmt.Println("Формат --range: start,end,step")
			return
		}
		start, _ := strconv.ParseFloat(parts[0], 64)
		end, _ := strconv.ParseFloat(parts[1], 64)
		step, _ := strconv.ParseFloat(parts[2], 64)
		if step <= 0 {
			fmt.Println("Шаг должен быть положительным")
			return
		}
		rows := []string{}
		for v := start; v <= end+1e-9; v += step {
			res, _ := convert(v, from, to)
			rows = append(rows, fmt.Sprintf("%.*f %s = %.*f %s",
				precision, v, scales[from].Symbol,
				precision, res, scales[to].Symbol))
		}
		if output != "" {
			os.WriteFile(output, []byte(strings.Join(rows, "\n")), 0644)
			fmt.Printf("Таблица сохранена в %s\n", output)
		} else {
			fmt.Printf("Таблица %s -> %s:\n", scales[from].Name, scales[to].Name)
			for _, row := range rows {
				fmt.Println(row)
			}
		}
		return
	}

	if batch != "" {
		file, err := os.Open(batch)
		if err != nil {
			fmt.Printf("Ошибка открытия %s: %v\n", batch, err)
			return
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		values := []float64{}
		for scanner.Scan() {
			line := strings.TrimSpace(scanner.Text())
			if line == "" {
				continue
			}
			v, err := strconv.ParseFloat(line, 64)
			if err != nil {
				fmt.Printf("Пропущено нечисловое значение: %s\n", line)
				continue
			}
			values = append(values, v)
		}
		results := []string{}
		for _, v := range values {
			res, _ := convert(v, from, to)
			results = append(results, fmt.Sprintf("%.*f -> %.*f", precision, v, precision, res))
		}
		if output != "" {
			os.WriteFile(output, []byte(strings.Join(results, "\n")), 0644)
			fmt.Printf("Результаты сохранены в %s\n", output)
		} else {
			for _, r := range results {
				fmt.Println(r)
			}
		}
		return
	}

	if flag.NFlag() == 0 {
		fmt.Println("Укажите --value или --batch или --range, или --list")
		return
	}

	res, err := convert(value, from, to)
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Printf("%.*f %s = %.*f %s\n", precision, value, scales[from].Symbol, precision, res, scales[to].Symbol)
}

File: temp_convert.rs
// temp_convert.rs - Конвертер температур на Rust (CLI)
// Зависимости: clap, colored
use clap::{Arg, App};
use colored::*;
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead, Write};
use std::str::FromStr;

lazy_static::lazy_static! {
    static ref SCALES: HashMap<String, Scale> = {
        let mut m = HashMap::new();
        m.insert("c".to_string(), Scale {
            to_kelvin: |x| x + 273.15,
            from_kelvin: |x| x - 273.15,
            name: "Цельсий".to_string(),
            symbol: "°C".to_string(),
        });
        // ... аналогично остальным шкалам (сокращённо для примера)
        m
    };
}

struct Scale {
    to_kelvin: fn(f64) -> f64,
    from_kelvin: fn(f64) -> f64,
    name: String,
    symbol: String,
}

fn convert(value: f64, from: &str, to: &str) -> Result<f64, String> {
    if from == to { return Ok(value); }
    let f = SCALES.get(from).ok_or_else(|| format!("Неизвестная шкала: {}", from))?;
    let t = SCALES.get(to).ok_or_else(|| format!("Неизвестная шкала: {}", to))?;
    let kelvin = (f.to_kelvin)(value);
    Ok((t.from_kelvin)(kelvin))
}

fn main() {
    let matches = App::new("Temp Converter")
        .arg(Arg::with_name("value").short("v").long("value").takes_value(true).help("Значение"))
        .arg(Arg::with_name("from").short("f").long("from").default_value("c").help("Исходная шкала"))
        .arg(Arg::with_name("to").short("t").long("to").default_value("f").help("Целевая шкала"))
        .arg(Arg::with_name("precision").short("p").long("precision").default_value("2").help("Точность"))
        .arg(Arg::with_name("batch").short("b").long("batch").takes_value(true).help("Файл со значениями"))
        .arg(Arg::with_name("output").short("o").long("output").takes_value(true).help("Выходной файл"))
        .arg(Arg::with_name("list").long("list").help("Показать шкалы"))
        .arg(Arg::with_name("range").long("range").takes_value(true).help("start,end,step"))
        .get_matches();

    if matches.is_present("list") {
        println!("Доступные шкалы:");
        for (key, s) in SCALES.iter() {
            println!("  {}: {} ({})", key, s.name, s.symbol);
        }
        return;
    }

    let precision: usize = matches.value_of("precision").unwrap().parse().unwrap_or(2);
    let from = matches.value_of("from").unwrap();
    let to = matches.value_of("to").unwrap();

    if let Some(range) = matches.value_of("range") {
        let parts: Vec<&str> = range.split(',').collect();
        if parts.len() != 3 {
            eprintln!("Формат --range: start,end,step");
            return;
        }
        let start: f64 = parts[0].parse().unwrap();
        let end: f64 = parts[1].parse().unwrap();
        let step: f64 = parts[2].parse().unwrap();
        if step <= 0.0 { eprintln!("Шаг должен быть положительным"); return; }
        let mut rows = Vec::new();
        let mut v = start;
        while v <= end + 1e-9 {
            let res = convert(v, from, to).unwrap();
            rows.push(format!("{:.prec$} {} = {:.prec$} {}", v, SCALES[from].symbol, res, SCALES[to].symbol, prec=precision));
            v += step;
        }
        if let Some(out) = matches.value_of("output") {
            let content = rows.join("\n");
            std::fs::write(out, content).unwrap();
            println!("Таблица сохранена в {}", out);
        } else {
            println!("Таблица {} -> {}:", SCALES[from].name, SCALES[to].name);
            for row in rows { println!("{}", row); }
        }
        return;
    }

    if let Some(batch_file) = matches.value_of("batch") {
        let file = File::open(batch_file).expect("Не удалось открыть файл");
        let reader = io::BufReader::new(file);
        let mut results = Vec::new();
        for line in reader.lines() {
            let line = line.unwrap();
            if line.trim().is_empty() { continue; }
            let val: f64 = line.trim().parse().unwrap_or(0.0);
            let res = convert(val, from, to).unwrap();
            results.push(format!("{:.prec$} -> {:.prec$}", val, res, prec=precision));
        }
        if let Some(out) = matches.value_of("output") {
            std::fs::write(out, results.join("\n")).unwrap();
            println!("Результаты сохранены в {}", out);
        } else {
            for r in results { println!("{}", r); }
        }
        return;
    }

    if let Some(val_str) = matches.value_of("value") {
        let value: f64 = val_str.parse().unwrap();
        let res = convert(value, from, to).unwrap();
        println!("{:.prec$} {} = {:.prec$} {}", value, SCALES[from].symbol, res, SCALES[to].symbol, prec=precision);
    } else {
        eprintln!("Укажите --value, --batch, --range или --list");
    }
}
